a3p-integration/proposals/z:acceptance/psm.test.js

```diff
@@ -0,0 +1,274 @@
+    anchor,
+    toIst,
+    govParams: { wantMintedFeeVal },
+  } = psmTestSpecs;
+
+  const psmTrader = await getUser(name);
+
+  const [metricsBefore, balances] = await Promise.all([
+    getPsmMetrics(anchor),
+    getBalances([psmTrader]),
+  ]);
+
+  const balancesBefore = await adjustBalancesIfNotProvisioned(
+    balances,
+    psmTrader,
+  );
+  t.log('METRICS', metricsBefore);
+  t.log('BALANCES', balancesBefore);
+
+  await agopsPsm(psmTrader, [
+    'swap',
+    '--pair',
+    process.env.PSM_PAIR,
+    '--wantMinted',
+    toIst.value,
+    '--feePct',
+    wantMintedFeeVal,
+  ]);
+  await waitForBlock(5);
+
+  await checkSwapSucceeded(t, metricsBefore, balancesBefore, {
+    wantMinted: toIst.value,
+    trader: psmTrader,
+    fee: Number(wantMintedFeeVal) / 100, // fee has to be between 0 and 1
+    anchor,
+  });
+});
+
+test.serial('swap out of IST', async t => {
+  const {
+    newUser: { name },
+    anchor,
+    fromIst,
+    govParams: { giveMintedFeeVal },
+  } = psmTestSpecs;
+
+  const psmTrader = await getUser(name);
+
+  const [metricsBefore, balancesBefore] = await Promise.all([
+    getPsmMetrics(anchor),
+    getBalances([psmTrader]),
+  ]);
+
+  t.log('METRICS', metricsBefore);
+  t.log('BALANCES', balancesBefore);
+
+  await agopsPsm(psmTrader, [
+    'swap',
+    '--pair',
+    process.env.PSM_PAIR,
+    '--giveMinted',
+    fromIst.value,
+    '--feePct',
+    giveMintedFeeVal,
+  ]);
+  await waitForBlock(5);
+
+  await checkSwapSucceeded(t, metricsBefore, balancesBefore, {
+    giveMinted: fromIst.value,
+    trader: psmTrader,
+    fee: Number(giveMintedFeeVal) / 100, // fee has to be between 0 and 1
+    anchor,
+  });
+});
+
+test.serial('mint limit is adhered', async t => {
+  const {
+    otherUser: {
+      fund: { denom, value },
+      name,
+    },
+    govParams,
+    anchor,
+  } = psmTestSpecs;
+
+  // Fund other user
+  const otherAddr = await getUser(name);
+  await bankSend(otherAddr, `${value}${denom}`);
+
+  // Replace when https://github.com/Agoric/agoric-sdk/pull/10171 is in
+  await waitForBlock(3);
+
+  const [metricsBefore, balancesBefore] = await Promise.all([
+    getPsmMetrics(anchor),
+    getBalances([otherAddr]),
+  ]);
+
+  t.log('METRICS', metricsBefore);
+  t.log('BALANCES', balancesBefore);
+
+  const { maxMintableValue, wantFeeValue } = await maxMintBelowLimit(anchor);
+  const maxMintFeesAccounted = Math.floor(
+    maxMintableValue * (1 - wantFeeValue),
+  );
+  t.log({ maxMintableValue, wantFeeValue, maxMintFeesAccounted });
+
+  // Send a swap, should fail because mint limit is exceeded
+  await agopsPsm(otherAddr, [
+    'swap',
+    '--pair',
+    process.env.PSM_PAIR,
+    '--wantMinted',
+    maxMintFeesAccounted / 1000000 + 2, // Make sure we exceed the limit
+    '--feePct',
+    govParams.wantMintedFeeVal,
+  ]);
+  await waitForBlock(5);
+
+  // Now check if failed with correct error message
+  await checkSwapExceedMintLimit(t, otherAddr, metricsBefore);
```

Comment:
Pushing the `t.is(...)` etc. down into `checkSwapExceedMintLimit` makes this code somewhat hard to follow.

What do you think of:

```suggestion
  const walletUpdate = await getWalletUpdate(otherAddr, 'offerStatus');
  t.is(walletUpdate.status.error, 'Error: Request would exceed mint limit');

  t.like(await getPsmMetrics(anchor), { mintedPoolBalance: metricsBefore.mintedPoolBalance });
```

where `getWalletUpdate` throws if `offerResult.updated != ` the 2nd arg.

Library functions that read data are straightforward. For non-read operations, making a swap is a clearly scoped thing to put in a helper function. But functions that mix reading data and doing assertions with `t` are hard to follow.

Reply:
ignored for now